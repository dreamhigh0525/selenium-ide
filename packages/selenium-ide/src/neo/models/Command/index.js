// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import { action, computed, observable, toJS } from 'mobx'
import uuidv4 from 'uuid/v4'
import Fuse from 'fuse.js'
import { Commands as _Commands, TargetTypes } from './Commands'
import { ArgTypes as _ArgTypes } from './ArgTypes'
const EventEmitter = require('events')
import { mergeEventEmitter } from '../../../common/events'

const DEFAULT_NEW_WINDOW_TIMEOUT = 2000
const EE = Symbol('event-emitter')

export default class Command {
  id = null
  @observable
  comment = ''
  @observable
  command
  @observable
  target
  @observable
  targets = []
  @observable
  value
  @observable
  isBreakpoint = false
  @observable
  opensWindow = false
  @observable
  windowHandleName = ''
  @observable
  windowTimeout = DEFAULT_NEW_WINDOW_TIMEOUT
  @observable
  opensWindowRead = false

  constructor(id = uuidv4(), command, target, value) {
    this.id = id
    this.command = command || ''
    this.target = target || ''
    this.value = value || ''
    this.export = this.export.bind(this)
    this[EE] = new EventEmitter()
    mergeEventEmitter(this, this[EE])
  }

  @computed
  get displayedName() {
    return this.enabled ? this.command : this.command.substr(2)
  }

  @computed
  get enabled() {
    return this.command && !this.command.startsWith('//')
  }

  @computed
  get isValid() {
    return Commands.array.includes(this.command)
  }

  @computed
  get canHaveTargets() {
    if (Commands.list.has(this.command)) {
      const { type } = Commands.list.get(this.command)
      return type === TargetTypes.LOCATOR
    }
    return false
  }

  @action.bound
  clone() {
    const clone = new Command()
    clone.setData(this.export())
    return clone
  }

  @action.bound
  setComment(comment) {
    this.comment = comment || ''
  }

  @action.bound
  setCommand(command) {
    if (Commands.values[command]) {
      this.command = Commands.values[command]
    } else {
      this.command = command || ''
    }

    if (!this.canHaveTargets) {
      this.setTargets()
    }

    if (this.command.indexOf('waitFor') > -1 && this.value === '') {
      this.setValue('30000')
    }
  }

  @action.bound
  setTarget(target) {
    this.target = target || ''
  }

  @action.bound
  setTargets(targets = []) {
    if (this.canHaveTargets) {
      this.targets.replace(targets)
    } else {
      this.targets.replace([])
    }
  }

  @action.bound
  setValue(value) {
    this.value = value ? value.replace(/\n/g, '\\n') : ''
  }

  @action.bound
  setOpensWindow(opensWindow) {
    if (typeof opensWindow == typeof true) {
      this.opensWindow = opensWindow
      if (!this.windowHandleName) {
        this.windowHandleName = `win${Math.floor(Math.random() * 10000)}`
      }
    }
  }

  @action.bound
  setWindowHandleName(newName) {
    const oldName = this.windowHandleName
    this.windowHandleName = newName
    this[EE].emit('window-handle-name-changed', oldName, newName)
  }

  @action.bound
  toggleOpensWindowRead() {
    this.opensWindowRead = !this.opensWindowRead
  }

  @action.bound
  setWindowTimeout(timeout) {
    this.windowTimeout = timeout
  }

  @action.bound
  toggleBreakpoint() {
    this.isBreakpoint = !this.isBreakpoint
  }

  @action.bound
  toggleEnabled() {
    if (this.enabled) {
      this.setCommand('//' + this.command)
    } else {
      this.setCommand(this.command.substr(2))
    }
  }

  @action.bound
  setData(jsRep) {
    this.setComment(jsRep.comment)
    this.setCommand(jsRep.command)
    this.setTarget(jsRep.target)
    this.setTargets(jsRep.targets)
    this.setValue(jsRep.value)

    if (jsRep.opensWindow) {
      this.setOpensWindow(jsRep.opensWindow)
      this.setWindowHandleName(jsRep.windowHandleName)
      this.setWindowTimeout(jsRep.windowTimeout)
    }
  }

  export() {
    let exported = {
      id: this.id,
      comment: this.comment,
      command: this.command,
      target: this.target,
      targets: toJS(this.targets),
      value: this.value,
    }

    if (this.opensWindow) {
      exported.opensWindow = this.opensWindow
      exported.windowHandleName = this.windowHandleName
      exported.windowTimeout = this.windowTimeout
    }

    return exported
  }

  static fromJS = function(jsRep) {
    const command = new Command(jsRep.id)
    command.setData(jsRep)
    return command
  }
}

export const ArgTypes = _ArgTypes

class CommandList {
  @observable
  list = new Map(_Commands)

  @computed
  get array() {
    return Array.from(this.list.keys())
  }

  @computed
  get values() {
    return this.array.reduce((commands, command) => {
      commands[this.list.get(command).name] = command
      return commands
    }, {})
  }

  @computed
  get fuse() {
    return new Fuse(Array.from(this.list.values()), {
      shouldSort: true,
      threshold: 0.4,
      location: 0,
      distance: 100,
      maxPatternLength: 50,
      minMatchCharLength: 1,
      keys: ['name'],
    })
  }

  @action.bound
  search(pattern) {
    return this.fuse.search(pattern)
  }

  @action.bound
  addCommand(id, name) {
    if (this.list.has(id)) {
      throw new Error(`Command with the id ${id} already exists`)
    } else {
      this.list.set(id, name)
    }
  }

  static get instance() {
    if (!this._instance) {
      this._instance = new CommandList()
    }
    return this._instance
  }
}

export const Commands = CommandList.instance

export const ControlFlowCommandNames = {
  do: 'do',
  else: 'else',
  elseIf: 'elseIf',
  end: 'end',
  forEach: 'forEach',
  if: 'if',
  repeatIf: 'repeatIf',
  times: 'times',
  while: 'while',
}

function commandNamesEqual(command, target) {
  if (command) {
    return command.command === target
  } else {
    return false
  }
}

function isBlockOpen(command) {
  return isIf(command) || isLoop(command)
}

function isConditional(command) {
  switch (command.command) {
    case ControlFlowCommandNames.elseIf:
    case ControlFlowCommandNames.if:
    case ControlFlowCommandNames.repeatIf:
    case ControlFlowCommandNames.times:
    case ControlFlowCommandNames.while:
      return true
    default:
      return false
  }
}

function isControlFlow(command) {
  switch (command.command) {
    case ControlFlowCommandNames.if:
    case ControlFlowCommandNames.elseIf:
    case ControlFlowCommandNames.else:
    case ControlFlowCommandNames.end:
    case ControlFlowCommandNames.do:
    case ControlFlowCommandNames.repeatIf:
    case ControlFlowCommandNames.times:
    case ControlFlowCommandNames.while:
      return true
    default:
      return false
  }
}

function isDo(command) {
  return commandNamesEqual(command, ControlFlowCommandNames.do)
}

function isElse(command) {
  return commandNamesEqual(command, ControlFlowCommandNames.else)
}

function isElseIf(command) {
  return commandNamesEqual(command, ControlFlowCommandNames.elseIf)
}

function isElseOrElseIf(command) {
  return isElseIf(command) || isElse(command)
}

function isEnd(command) {
  return commandNamesEqual(command, ControlFlowCommandNames.end)
}

function isIf(command) {
  return commandNamesEqual(command, ControlFlowCommandNames.if)
}

function isIfBlock(command) {
  return isIf(command) || isElseOrElseIf(command)
}

function isLoop(command) {
  return (
    commandNamesEqual(command, ControlFlowCommandNames.while) ||
    commandNamesEqual(command, ControlFlowCommandNames.times) ||
    commandNamesEqual(command, ControlFlowCommandNames.repeatIf) ||
    commandNamesEqual(command, ControlFlowCommandNames.forEach)
  )
}

function isTerminal(command) {
  return isElse(command) || isDo(command) || isEnd(command)
}

function isTimes(command) {
  return commandNamesEqual(command, ControlFlowCommandNames.times)
}

function isForEach(command) {
  return commandNamesEqual(command, ControlFlowCommandNames.forEach)
}

export const ControlFlowCommandChecks = {
  isIfBlock: isIfBlock,
  isConditional: isConditional,
  isDo: isDo,
  isElse: isElse,
  isElseOrElseIf: isElseOrElseIf,
  isEnd: isEnd,
  isIf: isIf,
  isLoop: isLoop,
  isBlockOpen: isBlockOpen,
  isTerminal: isTerminal,
  isControlFlow: isControlFlow,
  isTimes: isTimes,
  isForEach: isForEach,
}
