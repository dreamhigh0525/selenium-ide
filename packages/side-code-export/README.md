# SIDE Utils

Utils for Selenium IDE (e.g., code export)

## Installation

```js
npm install -g @seleniumhq/side-code-export
```

## Usage

```javascript
const { codeExport } = require('@seleniumhq/side-code-export')
```

__NOTE: For an example of usage, see one of the existing languages for code export in Selenium IDE (e.g., [`java-junit`](https://github.com/SeleniumHQ/selenium-ide/tree/trunk/packages/code-export-java-junit), [`javascript-mocha`](https://github.com/SeleniumHQ/selenium-ide/tree/trunk/packages/code-export-javascript-mocha), or [`python-pytest`](https://github.com/SeleniumHQ/selenium-ide/tree/trunk/packages/code-export-python-pytest)), ['csharp-xunit'](https://github.com/SeleniumHQ/selenium-ide/tree/trunk/packages/code-export-csharp-xunit).__
