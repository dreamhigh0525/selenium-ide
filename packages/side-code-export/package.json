{
  "name": "@seleniumhq/side-code-export",
  "version": "4.0.0-alpha.0",
  "description": "Utils for code export from Selenium IDE",
  "repository": "https://github.com/SeleniumHQ/selenium-ide",
  "keywords": [
    "selenium",
    "ide",
    "export"
  ],
  "license": "Apache-2.0",
  "private": false,
  "scripts": {
    "build": "babel -d dist src --ignore=__test__",
    "watch": "yarn build -w"
  },
  "main": "dist/index.js",
  "types": "dist/index.d.ts",
  "dependencies": {
    "@seleniumhq/side-model": "^4.0.0-alpha.0",
    "ua-parser-js": "0.7.28"
  },
  "gitHead": "f58e327e7616e23a3e926e4b80cf9952164e5744"
}
