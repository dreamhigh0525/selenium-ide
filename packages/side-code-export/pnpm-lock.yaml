lockfileVersion: 5.4

specifiers:
  '@seleniumhq/side-model': ^4.0.0-alpha.1
  ua-parser-js: 0.7.28

dependencies:
  '@seleniumhq/side-model': link:../side-model
  ua-parser-js: 0.7.28

packages:

  /ua-parser-js/0.7.28:
    resolution: {integrity: sha512-6Gurc1n//gjp9eQNXjD9O3M/sMwVtN5S8Lv9bvOYBfKfDNiIIhqiyi01vMBO45u4zkDE420w/e0se7Vs+sIg+g==}
    dev: false
