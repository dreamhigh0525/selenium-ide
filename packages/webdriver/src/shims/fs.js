const fs = require("fs");
fs.existsSync = (file) => {
  console.log("fs.existsSync: " + file);
  return false;
};

