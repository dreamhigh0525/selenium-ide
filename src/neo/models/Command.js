import { action, computed, observable } from "mobx";
import uuidv4 from "uuid/v4";

export default class Command {
  id = null;
  @observable command = "";
  @observable target = "";
  @observable value = "";

  constructor(id = uuidv4()) {
    this.id = id;
    this.setCommand = this.setCommand.bind(this);
    this.setTarget = this.setTarget.bind(this);
    this.setValue = this.setValue.bind(this);
  }

  @action setCommand(command) {
    if (CommandsValues[command]) {
      this.command = CommandsValues[command];
    } else {
      this.command = command;
    }
  }

  @computed get isValid() {
    return CommandsArray.includes(this.command);
  }

  @action setTarget(target) {
    this.target = target;
  }

  @action setValue(value) {
    this.value = value;
  }
}

export const Commands = Object.freeze({
  addSelection: "add selection",
  answerOnNextPrompt: "answer on next prompt",
  assertAlert: "assert alert",
  assertConfirmation: "assert confirmation",
  assertPrompt: "assert prompt",
  assertText: "assert text",
  assertTitle: "assert title",
  chooseCancelOnNextConfirmation: "choose cancel on next confirmation",
  chooseCancelOnNextPrompt: "choose cancel on next prompt",
  chooseOkOnNextConfirmation: "choose ok on next confirmation",
  clickAt: "click at",
  doubleClickAt: "double click at",
  dragAndDropToObject: "drag and drop to object",
  echo: "echo",
  editContent: "edit content",
  mouseDownAt: "mouse down at",
  mouseMoveAt: "mouse move at",
  mouseOut: "mouse out",
  mouseOver: "mouse over",
  mouseUpAt: "mouse up at",
  open: "open",
  pause: "pause",
  removeSelection: "remove selection",
  runScript: "run script",
  select: "select",
  selectFrame: "select frame",
  selectWindow: "select window",
  sendKeys: "send keys",
  store: "store",
  storeText: "store text",
  storeTitle: "store title",
  type: "type",
  verifyText: "verify text",
  verifyTitle: "verify title"
});

export const CommandsArray = Object.freeze(Object.keys(Commands));

export const CommandsValues = Object.freeze(CommandsArray.reduce((commands, command) => {
  commands[Commands[command]] = command;
  return commands;
}, {}));
