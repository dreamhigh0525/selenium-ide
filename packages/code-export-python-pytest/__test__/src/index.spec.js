// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

import fs from 'fs'
import path from 'path'
import codeExport, {
  emitTest,
  emitSuite,
  _emitMethod,
  _findTestByName,
} from '../../src'
import Command from '../../src/command'
import { project as projectProcessor } from '@seleniumhq/side-code-export'

function readFile(filename) {
  return JSON.parse(
    fs.readFileSync(
      path.join(
        __dirname,
        '..',
        '..',
        '..',
        'side-code-export',
        '__test__',
        'test-files',
        filename
      )
    )
  )
}

describe('Code Export Python pytest', () => {
  it('should export a test', async () => {
    const project = readFile('single-test.side')
    const results = await emitTest({
      baseUrl: project.url,
      test: project.tests[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a test with grid execution', async () => {
    const project = readFile('single-test.side')
    const results = await emitTest({
      baseUrl: project.url,
      test: project.tests[0],
      tests: project.tests,
      beforeEachOptions: {
        browserName: 'Firefox',
        gridUrl: 'http://localhost:4444/wd/hub',
      },
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a suite', async () => {
    const project = normalizeProject(readFile('single-suite.side'))
    const results = await emitSuite({
      baseUrl: project.url,
      suite: project.suites[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a test with a reused test method', async () => {
    const project = normalizeProject(readFile('test-case-reuse.side'))
    const results = await emitTest({
      baseUrl: project.url,
      test: project.tests[1],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a suite with a reused test method', async () => {
    const project = normalizeProject(readFile('test-case-reuse.side'))
    const results = await emitSuite({
      baseUrl: project.url,
      suite: project.suites[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a suite that uses control flow commands', async () => {
    const project = normalizeProject(readFile('control-flow-suite.side'))
    const results = await emitSuite({
      baseUrl: project.url,
      suite: project.suites[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a test with commands that open a new window', async () => {
    const project = normalizeProject(readFile('select-window.side'))
    const results = await emitTest({
      baseUrl: project.url,
      test: project.tests[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a suite with commands that open a new window inside of a reused test method', async () => {
    const project = normalizeProject(readFile('nested-select-window.side'))
    const results = await emitSuite({
      baseUrl: project.url,
      suite: project.suites[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export a suite with just one new window util method when there are multiple commands that open a new window', async () => {
    const project = normalizeProject(readFile('nested-select-window-v2.side'))
    const results = await emitSuite({
      baseUrl: project.url,
      suite: project.suites[0],
      tests: project.tests,
    })
    expect(results.body).toBeDefined()
    expect(results.body).toMatchSnapshot()
  })
  it('should export the command emitter', () => {
    expect(Command.emit).toStrictEqual(codeExport.emit.command)
  })
})

function normalizeProject(project) {
  let _project = { ...project }
  _project.suites.forEach(suite => {
    projectProcessor.normalizeTestsInSuite({ suite, tests: _project.tests })
  })
  return _project
}
