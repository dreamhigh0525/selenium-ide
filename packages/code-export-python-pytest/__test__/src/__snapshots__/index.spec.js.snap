// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Code Export Java JUnit Selenium should export a suite 1`] = `
"# Generated by Selenium IDE
import pytest
import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait

class TestLogin():
  def setup_method(self, method):
    self.driver = webdriver.Firefox()
    self.vars = {}
  
  def teardown_method(self, method):
    self.driver.quit();
  
  def test_validcredentials(self):
    self.driver.get(\\"http://the-internet.herokuapp.com/login\\")
    self.driver.set_window_size(1440, 1177)
    self.driver.find_element(By.ID, \\"username\\").send_keys(\\"tomsmith\\")
    self.driver.find_element(By.ID, \\"password\\").send_keys(\\"SuperSecretPassword!\\")
    self.driver.find_element(By.CSS_SELECTOR, \\"#login button\\").click()
    assert self.driver.find_element(By.CSS_SELECTOR, \\".flash.success\\").text == \\"You logged into a secure area!\\\\\\\\n×\\"
  
  def test_invalidcredentials(self):
    self.driver.get(\\"http://the-internet.herokuapp.com/login\\")
    self.driver.set_window_size(1440, 1177)
    self.driver.find_element(By.ID, \\"username\\").send_keys(\\"blah\\")
    self.driver.find_element(By.ID, \\"password\\").send_keys(\\"blah\\")
    self.driver.find_element(By.CSS_SELECTOR, \\"#login button\\").click()
    assert self.driver.find_element(By.ID, \\"flash\\").text == \\"Your username is invalid!\\\\\\\\n×\\"
  
"
`;

exports[`Code Export Java JUnit Selenium should export a suite that uses control flow commands 1`] = `
"# Generated by Selenium IDE
import pytest
import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait

class TestControlflow():
  def setup_method(self, method):
    self.driver = webdriver.Firefox()
    self.vars = {}
  
  def teardown_method(self, method):
    self.driver.quit();
  
  def test_controlflowif(self):
    vars[\\"myVar\\"] = self.driver.execute_script(\\"return 'a'\\")
    if self.driver.execute_script(\\"return (arguments[0] === 'a')\\", vars[\\"myVar\\"]):
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'a'\\")
    elif self.driver.execute_script(\\"return (arguments[0] === 'b')\\", vars[\\"myVar\\"]):
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'b'\\")
    else:
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'c'\\")
    assert(vars[\\"output\\"] == \\"a\\")
  
  def test_controlflowelseif(self):
    vars[\\"myVar\\"] = self.driver.execute_script(\\"return 'b'\\")
    if self.driver.execute_script(\\"return (arguments[0] === 'a')\\", vars[\\"myVar\\"]):
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'a'\\")
    elif self.driver.execute_script(\\"return (arguments[0] === 'b')\\", vars[\\"myVar\\"]):
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'b'\\")
    else:
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'c'\\")
    assert(vars[\\"output\\"] == \\"b\\")
  
  def test_controlflowelse(self):
    vars[\\"myVar\\"] = self.driver.execute_script(\\"return 'c'\\")
    if self.driver.execute_script(\\"return (arguments[0] === 'a')\\", vars[\\"myVar\\"]):
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'a'\\")
    elif self.driver.execute_script(\\"return (arguments[0] === 'b')\\", vars[\\"myVar\\"]):
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'b'\\")
    else:
      vars[\\"output\\"] = self.driver.execute_script(\\"return 'c'\\")
    assert(vars[\\"output\\"] == \\"c\\")
  
  def test_controlflowdo(self):
    vars[\\"check\\"] = self.driver.execute_script(\\"return 1\\")
    condition = True
    while condition:
      vars[\\"check\\"] = self.driver.execute_script(\\"return arguments[0] + 1\\", vars[\\"check\\"])
    condition = self.driver.execute_script(\\"return (arguments[0] < 3)\\", vars[\\"check\\"])
    assert(vars[\\"check\\"] == \\"3\\")
  
  def test_controlflowtimes(self):
    vars[\\"check\\"] = self.driver.execute_script(\\"return 1\\")
    for i in range(1, 2):
      vars[\\"check\\"] = self.driver.execute_script(\\"return arguments[0] + 1\\", vars[\\"check\\"])
    assert(vars[\\"check\\"] == \\"3\\")
  
  def test_controlflowwhile(self):
    vars[\\"check\\"] = self.driver.execute_script(\\"return 1\\")
    while self.driver.execute_script(\\"return (arguments[0] < 3)\\", vars[\\"check\\"]):
      vars[\\"check\\"] = self.driver.execute_script(\\"return arguments[0] + 1\\", vars[\\"check\\"])
    assert(vars[\\"check\\"] == \\"3\\")
  
"
`;

exports[`Code Export Java JUnit Selenium should export a suite with a reused test method 1`] = `
"# Generated by Selenium IDE
import pytest
import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait

class TestLogin():
  def setup_method(self, method):
    self.driver = webdriver.Firefox()
    self.vars = {}
  
  def teardown_method(self, method):
    self.driver.quit();
  
  def login(self):
    self.driver.get(\\"https://the-internet.herokuapp.com/\\")
    self.driver.find_element(By.LINK_TEXT, \\"Form Authentication\\").click()
    self.driver.find_element(By.ID, \\"username\\").send_keys(vars[\\"username\\"])
    self.driver.find_element(By.ID, \\"password\\").send_keys(vars[\\"password\\"])
    self.driver.find_element(By.CSS_SELECTOR, \\".fa\\").click()
  
  def test_invalidcredentials(self):
    vars[\\"username\\"] = \\"blah\\"
    vars[\\"password\\"] = \\"blah\\"
    self.login()
  
  def test_validcredentials(self):
    vars[\\"username\\"] = \\"tomsmith\\"
    vars[\\"password\\"] = \\"SuperSecretPassword!\\"
    self.login()
  
"
`;

exports[`Code Export Java JUnit Selenium should export a suite with commands that open a new window inside of a reused test method 1`] = `
"# Generated by Selenium IDE
import pytest
import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait

class TestDefaultSuite():
  def setup_method(self, method):
    self.driver = webdriver.Firefox()
    self.vars = {}
  
  def teardown_method(self, method):
    self.driver.quit();
  
  def wait_for_window(timeout = 2000) {
    self.driver.sleep(timeout)
    wh_now = self.driver.window_handles
    wh_then = vars[\\"window_handles\\"]
    if len(wh_now) > len(wh_then):
      return set(wh_now).difference(set(wh_then))
  
  def selectwindow(self):
    self.driver.get(\\"http://the-internet.herokuapp.com/windows\\")
    vars[\\"window_handles\\"] = self.driver.window_handles
    self.driver.find_element(By.LINK_TEXT, \\"Click Here\\").click()
    vars[\\"win9425\\"] = self.wait_for_window(2000)
    self.driver.switch_to.window(vars[\\"win9425\\"])
    assert self.driver.title == \\"New Window\\"
  
  def test_reuse(self):
    self.selectwindow()
  
"
`;

exports[`Code Export Java JUnit Selenium should export a suite with just one new window util method when there are multiple commands that open a new window 1`] = `
"# Generated by Selenium IDE
import pytest
import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait

class TestDefaultSuite():
  def setup_method(self, method):
    self.driver = webdriver.Firefox()
    self.vars = {}
  
  def teardown_method(self, method):
    self.driver.quit();
  
  def wait_for_window(timeout = 2000) {
    self.driver.sleep(timeout)
    wh_now = self.driver.window_handles
    wh_then = vars[\\"window_handles\\"]
    if len(wh_now) > len(wh_then):
      return set(wh_now).difference(set(wh_then))
  
  def selectwindow(self):
    self.driver.get(\\"http://the-internet.herokuapp.com/windows\\")
    vars[\\"window_handles\\"] = self.driver.window_handles
    self.driver.find_element(By.LINK_TEXT, \\"Click Here\\").click()
    vars[\\"win9425\\"] = self.wait_for_window(2000)
    self.driver.switch_to.window(vars[\\"win9425\\"])
    assert self.driver.title == \\"New Window\\"
  
  def test_reuse(self):
    self.selectwindow()
  
  def test_selectwindow(self):
    self.driver.get(\\"http://the-internet.herokuapp.com/windows\\")
    vars[\\"window_handles\\"] = self.driver.window_handles
    self.driver.find_element(By.LINK_TEXT, \\"Click Here\\").click()
    vars[\\"win9425\\"] = self.wait_for_window(2000)
    self.driver.switch_to.window(vars[\\"win9425\\"])
    assert self.driver.title == \\"New Window\\"
  
"
`;

exports[`Code Export Java JUnit Selenium should export a test 1`] = `
"# Generated by Selenium IDE
import pytest
import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait

class TestLogin():
  def setup_method(self, method):
    self.driver = webdriver.Firefox()
    self.vars = {}
  
  def teardown_method(self, method):
    self.driver.quit();
  
  def test_login(self):
    self.driver.get(\\"http://the-internet.herokuapp.com/login\\")
    self.driver.set_window_size(1440, 1177)
    self.driver.find_element(By.ID, \\"username\\").send_keys(\\"tomsmith\\")
    self.driver.find_element(By.ID, \\"password\\").send_keys(\\"SuperSecretPassword!\\")
    self.driver.find_element(By.CSS_SELECTOR, \\"#login button\\").click()
    assert self.driver.find_element(By.CSS_SELECTOR, \\".flash.success\\").text == \\"You logged into a secure area!\\\\\\\\n×\\"
  
"
`;

exports[`Code Export Java JUnit Selenium should export a test with a reused test method 1`] = `
"# Generated by Selenium IDE
import pytest
import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait

class TestValidcredentials():
  def setup_method(self, method):
    self.driver = webdriver.Firefox()
    self.vars = {}
  
  def teardown_method(self, method):
    self.driver.quit();
  
  def login(self):
    self.driver.get(\\"https://the-internet.herokuapp.com/\\")
    self.driver.find_element(By.LINK_TEXT, \\"Form Authentication\\").click()
    self.driver.find_element(By.ID, \\"username\\").send_keys(vars[\\"username\\"])
    self.driver.find_element(By.ID, \\"password\\").send_keys(vars[\\"password\\"])
    self.driver.find_element(By.CSS_SELECTOR, \\".fa\\").click()
  
  def test_validcredentials(self):
    vars[\\"username\\"] = \\"tomsmith\\"
    vars[\\"password\\"] = \\"SuperSecretPassword!\\"
    self.login()
  
"
`;

exports[`Code Export Java JUnit Selenium should export a test with commands that open a new window 1`] = `
"# Generated by Selenium IDE
import pytest
import time
import json
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.support.wait import WebDriverWait

class TestSelectwindow():
  def setup_method(self, method):
    self.driver = webdriver.Firefox()
    self.vars = {}
  
  def teardown_method(self, method):
    self.driver.quit();
  
  def wait_for_window(timeout = 2000) {
    self.driver.sleep(timeout)
    wh_now = self.driver.window_handles
    wh_then = vars[\\"window_handles\\"]
    if len(wh_now) > len(wh_then):
      return set(wh_now).difference(set(wh_then))
  
  def test_selectwindow(self):
    self.driver.get(\\"http://the-internet.herokuapp.com/\\")
    self.driver.set_window_size(1260, 1027)
    self.driver.find_element(By.LINK_TEXT, \\"Multiple Windows\\").click()
    vars[\\"window_handles\\"] = self.driver.window_handles
    self.driver.find_element(By.LINK_TEXT, \\"Click Here\\").click()
    vars[\\"win9625\\"] = self.wait_for_window(2000)
    vars[\\"root\\"] = self.driver.current_window_handle
    self.driver.switch_to.window(vars[\\"win9625\\"])
    assert self.driver.title == \\"New Window\\"
    self.driver.switch_to.window(vars[\\"root\\"])
    assert self.driver.title == \\"The Internet\\"
  
"
`;
